### PR TITLE
Attachments for lib/pure/smtp

### PR DESCRIPTION
**Pull request by enthus1ast**
Added simple attachments to lib/pure/smtp

### CODE REVIEW

**Comment by dom96** [comment]
Thank you for this PR! It's your first right? I was reading the IRC logs and seeing that you were a bit nervous about creating it. Please never feel nervous about creating PRs, especially to the Nim repo! :)

These kinds of PRs really help us so thank you again for taking the time to implement it and test it. There are some things that I would like for you to change before merging this PR, I will go through the diff now and make comments inline. If you don't have the time to make those changes then please let me know, I can do them or ask someone else to. More importantly please don't feel bad about the changes I ask you to make, many of them will be quite simple (style changes) and are just a result of me being very picky :)

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,7 @@
 *
 !**/
 !*.*
-nimcache/
+**nimcache**
```
Comment (dom96) [inline .gitignore]: This change doesn't relate to the purpose of this PR, please remove these changes and make them in a different branch then create a separate PR for them so that they can be discussed separately.

File: lib/pure/ospaths.nim [hunk 2]
```diff
@@ -383,6 +383,13 @@ when not declared(getEnv) or defined(nimscript):
     else:
       result = splitPath(path).tail
 
+  proc extractExt*(path: string): string {.
```
Comment (dom96) [inline lib/pure/ospaths.nim]: This procedure is not necessary. The ``os`` module defines a ``splitFile`` procedure which splits a filename for you into ``dir``, ``name`` and ``ext``, you can simply write ``splitFile(path).ext``.

File: lib/pure/smtp.nim [hunk 3]
```diff
@@ -28,20 +28,32 @@
 ## For SSL support this module relies on OpenSSL. If you want to
 ## enable SSL, compile with ``-d:ssl``.
 
-import net, strutils, strtabs, base64, os
+import net, strutils, strtabs, base64, os, mimetypes, sequtils
 import asyncnet, asyncdispatch
+import mersenne , times ## we need randomness for the boundary
```
Comment (dom96) [inline lib/pure/smtp.nim]: Please keep the style consistent, in this case there should be no space before the comma.

File: lib/pure/smtp.nim [hunk 4]
```diff
@@ -28,20 +28,32 @@
...
 
 type
   Smtp* = object
     sock: Socket
     debug: bool
 
+  Attachment* = object ## this is a mail attachment
+    mimeType: string ## the mime type of our attachement
+    filename: string ## the user spezified name of the attached file
```
Comment (dom96) [inline lib/pure/smtp.nim]: *specified

File: lib/pure/smtp.nim [hunk 5]
```diff
@@ -28,20 +28,32 @@
...
+    contentB64 : string ## the actual content of the file base 64ed
+    #description : string ## a description of this attachement
```
Comment (dom96) [inline lib/pure/smtp.nim]: Implement the ``description`` or remove the field completely.

File: lib/pure/smtp.nim [hunk 6]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+proc getMimeTypeOfFile(pathToFile:string): string =
+  ## This will return the mimetype
+  var mimeDB = newMimetypes()
+  var ext = extractExt(pathToFile) # 
+
+  ## we choose "application/octet-stream" 
+  ## if no extension is given
+  ## TODO what about linux? or files without extension.
+  ## Do we have to have something like `file` magic?
+  return mimeDB.getMimetype(ext,"bin")
```
Comment (dom96) [inline lib/pure/smtp.nim]: I think this is good as a default heuristic for determining the mimetype. What I would like you to do is to overload the ``attach`` procedure so that the user can optionally specify the mimetype, something like this:

```nim
proc attach*(msg: var Message, filepath: string, msgFilename: string) = ...
proc attach*(msg: var Message, filepath: string, msgFilename: string, mimetype: string) = ...
```

File: lib/pure/smtp.nim [hunk 7]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+
+  ## we choose "application/octet-stream" 
+  ## if no extension is given
+  ## TODO what about linux? or files without extension.
+  ## Do we have to have something like `file` magic?
+  return mimeDB.getMimetype(ext,"bin") 
+
+
+
+proc composeAttachement(attachement:Attachment,boundary:string) : string=
```
Comment (dom96) [inline lib/pure/smtp.nim]: Should be spelled ``attachment``.

File: lib/pure/smtp.nim [hunk 8]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+  ## this generates an attachement
+  result =""
+
+  var ContentDisposition :string = "Content-Disposition: attachment;"
```
Comment (dom96) [inline lib/pure/smtp.nim]: variable names shouldn't start with an upper case letter.

File: lib/pure/smtp.nim [hunk 9]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+  if attachement.filename != nil:
+    ContentDisposition &= " filename=" & attachement.filename & ";"
+
+  result.add("\c\L")
+  result.add("Content-Type: " & attachement.mimeType & "\c\L")
+  result.add( ContentDisposition & "\c\L")
```
Comment (dom96) [inline lib/pure/smtp.nim]: Please remove spaces such as these (i.e. after the open parenthesis).

File: lib/pure/smtp.nim [hunk 10]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+  result.add("MIME-Version: 1.0\c\L")
+  result.add("Content-Transfer-Encoding: base64 \c\L")
+  result.add("" & "\c\L")
+  result.add(attachement.contentB64 & "\c\L")
+  result.add("\c\L")
+  result.add("--" & boundary )
```
Comment (dom96) [inline lib/pure/smtp.nim]: ... and also before the closed parenthesis.

File: lib/pure/smtp.nim [hunk 11]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+
+
+proc attach*(msg:var Message,pathToFile:string,
```
Comment (dom96) [inline lib/pure/smtp.nim]: Rename the parameters of this procedure to ``msg, filepath, msgFilename``

File: lib/pure/smtp.nim [hunk 12]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+  attachement.mimeType    = getMimeTypeOfFile(pathToFile)
+  let content             = readFile(pathToFile)
+  attachement.contentB64  = encode(content)
+  attachement.filename    = myFilename
+  #attachement.description = description
+  add(msg.attachements,attachement)
+
+
+proc getRandomBoundary():string = 
+  var mt = newMersenneTwister( int( epochTime() )  ) #we dont need stong random
```
Comment (dom96) [inline lib/pure/smtp.nim]: Please use ``math.random`` instead.

File: lib/pure/smtp.nim [hunk 13]
```diff
@@ -141,31 +153,92 @@ proc close*(smtp: Smtp) =
...
+  return $mt.getNum()
+
+
+proc createMessage*(mSubject, mBody: string, mFrom:string,
```
Comment (dom96) [inline lib/pure/smtp.nim]: Why is the ``mFrom`` required? Adding this parameter will cause existing code to break.

**Comment by dom96** [comment]
General observations:

* Please keep the style consistent, make sure that commas ``,`` have a space after but no space before, same goes for colons (``:``). The equals sign (``=``) should have a space both before and after. When starting a comment add a space after the ``#``.

File: lib/pure/smtp.nim [hunk 14]
```diff
@@ -174,7 +247,22 @@ proc `$`*(msg: Message): string =
...
+  result.add("\c\L")
   result.add(msg.msgBody)
+  result.add("\c\L")
+
+  if msg.attachements.len() > 0: # if we have some attachements
+    result.add("\c\L")
+    result.add("--" & msg.boundary) # we start with a boundary
+
+    for attachement in msg.attachements:
+      # echo "FOOOOO"
```
Comment (dom96) [inline lib/pure/smtp.nim]: Simply remove testing code like this instead of commenting it out.

File: lib/pure/smtp.nim [hunk 15]
```diff
@@ -221,20 +309,22 @@ proc auth*(smtp: AsyncSmtp, username, password: string) {.async.} =
   ## May fail with ReplyError.
 
   await smtp.sock.send("AUTH LOGIN\c\L")
-  await smtp.checkReply("334") # TODO: Check whether it's asking for the "Username:"
+  await smtp.checkReply("334") # TODO: Check whether it's asking for the 
+                               # "Username:"
```
Comment (dom96) [inline lib/pure/smtp.nim]: Why did you move part of this comment onto the next line? (and the ones below too?)

**Comment by dom96** [comment]
I hope this doesn't discourage you. Feel free to ask for clarifications on any of the points that I made.

File: lib/pure/smtp.nim (hunk 13 above)
Comment (enthus1ast) [inline lib/pure/smtp.nim]: some smtp servers (eg: gmail) will not accept a message without a from field.
This fixes a bug in the old implementation.

**Comment by enthus1ast** [comment]
I'll correct the mistakes later that day.

File: lib/pure/smtp.nim (hunk 13 above)
Comment (dom96) [inline lib/pure/smtp.nim]: Ahh. Could you define the old procedure (without the ``mFrom`` param) but make it deprecated and explain why in the doc comment for that procedure?

**Comment by dom96** [comment]
@enthus1ast Thank you! :)

**Comment by cheatfate** [comment]
@dom96 i think it would be better to leave **smtp.nim** as transport module, and create new module **mime.nim** where this code can be placed, because attachments (belongs to MIME) are mainly contain procedures working with text (encoding, parsing). 
Another reason to place this code in another module, is that it is used not only in the SMTP protocol but in HTTP protocol too (POST multipart/form-data).

**Comment by dom96** [comment]
@cheatfate which code? There is very little mime-specific code in this as far as I can see.

**Comment by cheatfate** [comment]
@dom96 this code with very little mime-specific and it can be start of big mime.nim

**Comment by dom96** [comment]
@cheatfate we already have http://nim-lang.org/docs/mimetypes.html

**Comment by cheatfate** [comment]
@dom96, please just look [https://en.wikipedia.org/wiki/MIME](https://en.wikipedia.org/wiki/MIME).
> Although MIME was designed mainly for SMTP, the content types defined by MIME standards are also of importance outside of email, such as in communication protocols such as HTTP for the World Wide Web.
We already have this code in our standart library in **httpclient.nim** (Lines: 280-381).
Or you want to implement MIME specific code for every internet protocol it uses?

**Comment by dom96** [comment]
@cheatfate you're right, there are similarities but that code is already in ``httpclient`` and its specific to that module. I've already asked @enthus1ast for a lot of changes. Let's wait until those are fixed, then we can worry about this problem.

**Comment by enthus1ast** [comment]
@cheatfate this looks very nice, i've missed that one...

**Comment by enthus1ast** [comment]
I could port smtp to use the procs from httpclient.
But as far as i see, there would be also some patches to httpclient necessary.

**Comment by cheatfate** [comment]
@enthus1ast its why i want your code in different module (something like **mime.nim**), so we can use it in **httpclient.nim** and for sending emails with attachments.

**Comment by enthus1ast** [comment]
i think then its best to cancel this pull request for now right?

**Comment by cheatfate** [comment]
@enthus1ast I can not take any decisions, I can only propose or advise.
